### PR TITLE
Attach autograd edges only for tensors requiring grad.

### PR DESCRIPTION
**Pull request by pritamdamania87**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#30811 Attach autograd edges only for tensors requiring grad.**

When we sent tensors over RPC, on the server side we would call
addRecvRpcBackward which would call `set_history` on all tensors. This was
incorrect and set the `requires_grad` flag on tensors that didn't actually need
grad.

To fix this, we only attach autograd edges to tensors that need grads.

Differential Revision: [D18828561](https://our.internmc.facebook.com/intern/diff/D18828561/)

### CODE REVIEW

**Comment by pritamdamania87** [comment]
@zzzwen @mrshenli Haven't found the root cause of the incorrect gradient issue (will continue to debug). A simple test in this PR validates that the gradients are fine for a simple EmbeddingBag call.

**Review by mrshenli — APPROVED** [review]
This LGTM. Let's also add one non-EmbeddingBag example to test general use cases. Please feel free to copy the test from #30792

File: torch/csrc/distributed/autograd/utils.cpp [hunk 1]
```diff
@@ -23,12 +23,21 @@ void addSendRpcBackward(
     const AutogradMetadata& autogradMetadata,
     std::vector<torch::Tensor>& tensors,
     const rpc::worker_id_t dst) {
+  // Attach autograd information only for tensors requiring grad.
+  std::vector<torch::Tensor> tensors_with_grad;
+  std::copy_if(
```
Comment (zzzwen) [inline torch/csrc/distributed/autograd/utils.cpp]: Question, does`copy_if` create a deepcopy of the tensor?  It might not be idea to make a copy of  the tensors. 
I did not find a `move_if` in std. if we do not need to original vector, we might want to use `remove_if` from the original Tensor.

**Comment by zzzwen** [comment]
As for the difference between local and dist_autograd, the test I have is using RowWiseAdagrad.  @wanchaol, Do you know if there is a smaller test (does not include the training toolkit stack) for RowWiseAdagrad with EmbeddingBags,

File: torch/csrc/distributed/autograd/utils.cpp (hunk 1 above)
Comment (pritamdamania87) [reply torch/csrc/distributed/autograd/utils.cpp]: @zzzwen Tensor internally has a reference counted pointer, only that pointer is copied. This is not a deep copy of the tensor.

**Comment by pritamdamania87** [comment]
Abandoning in favor of: https://github.com/pytorch/pytorch/pull/30904
